### PR TITLE
feat: implement SwitchClause

### LINKED ISSUE #105: Support `if` clauses in `depends_components` and `depends_filepatterns`
For example

```yaml
folder_foo:
  depends_components:
    - if: IDF_VERSION == "5.2.0":
      components: ["esp_eth"]
    - if: IDF_VERSION == "5.3.0":
      components: ["esp_wifi"]
    - components: ["esp_eth", "esp_wifi"]
```

To keep the similarity with others, here the folder rule should also apply the first matched `if` statement. but on extra, we support one more "default" switch case.

explanation:

- if idf 5.2.0, depends_components: esp_eth
- if idf 5.3.0, depends_components: esp_wifi
- by default (other cases), depends_components: esp_eth, esp_wifi

`depends_filepatterns` should be something alike.

### PR DESCRIPTION
**Pull request by horw**
Implemented SwitchClause to enable the ability to use the following manifest format:
```
folder_foo:
  depends_components:
    - if: IDF_VERSION == "5.2.0":
      content: ["esp_eth"]
    - if: IDF_VERSION == "5.3.0":
      content: ["esp_wifi"]
    - default: ["esp_eth", "esp_wifi"]
```
or
```
folder_foo:
  depends_components:
    - esp_eth_1
    - esp_eth_2
```

closes #105

### CODE REVIEW

**Comment by horw** [comment]
@hfudev Hi! I have updated the documentation for the switch statement. Please take a look

File: idf_build_apps/manifest/manifest.py [hunk 1]
```diff
@@ -68,11 +81,43 @@ def _clause_to_if_clause(clause: t.Dict[str, t.Any]) -> IfClause:
...
+
+            return SwitchClause(ifs, values, default)
+
         self.enable = [_clause_to_if_clause(clause) for clause in enable] if enable else []
         self.disable = [_clause_to_if_clause(clause) for clause in disable] if disable else []
         self.disable_test = [_clause_to_if_clause(clause) for clause in disable_test] if disable_test else []
-        self.depends_components = depends_components or []
-        self.depends_filepatterns = depends_filepatterns or []
+        self.depends_components = _clauses_to_switch_clause(depends_components)
+        self.depends_filepatterns = _clauses_to_switch_clause(depends_filepatterns)
```
Comment (hfudev) [inline idf_build_apps/manifest/manifest.py:120]: I'm not sure if it's a breaking change or not.

I recommend to keep the original data type. i.e.

- list of dictionary -> `SwitchClause`
- list of string -> `list[str]`
- None -> `[]`

File: idf_build_apps/manifest/manifest.py [hunk 2]
```diff
@@ -232,8 +277,16 @@ def enable_test_targets(
     ) -> t.List[str]:
         return self._most_suitable_rule(folder).enable_test_targets(default_sdkconfig_target, config_name)
 
-    def depends_components(self, folder: str) -> t.List[str]:
-        return self._most_suitable_rule(folder).depends_components
+    def depends_components(
+        self, folder: str, default_sdkconfig_target: t.Optional[str] = None, config_name: t.Optional[str] = None
```
Comment (hfudev) [inline idf_build_apps/manifest/manifest.py:281]:
```suggestion
        self, folder: str, target: t.Optional[str] = None, config_name: t.Optional[str] = None
```

to keep the naming consistency with others

File: idf_build_apps/manifest/manifest.py [hunk 3]
```diff
@@ -232,8 +277,16 @@ def enable_test_targets(
...
+    ) -> t.List[str]:
+        return self._most_suitable_rule(folder).depends_components.get_value(
+            default_sdkconfig_target or '', config_name or ''
+        )
 
-    def depends_filepatterns(self, folder: str) -> t.List[str]:
-        return self._most_suitable_rule(folder).depends_filepatterns
+    def depends_filepatterns(
+        self, folder: str, default_sdkconfig_target: t.Optional[str] = None, config_name: t.Optional[str] = None
```
Comment (hfudev) [inline idf_build_apps/manifest/manifest.py:288]:
```suggestion
        self, folder: str, target: t.Optional[str] = None, config_name: t.Optional[str] = None
```

same

**Review by hfudev — CHANGES_REQUESTED** [review]
LGTM! Thanks for this nice feature.

left a few comments, and could you help add some tests on combining the `depends_components+` together with the switch clauses?

File: idf_build_apps/manifest/manifest.py (hunk 2 above)
Comment (horw) [reply idf_build_apps/manifest/manifest.py:281]: Agree, that's why I chose to use the name "default_sdkconfig_target."
```python
 def enable_build_targets(
        self, folder: str, default_sdkconfig_target: t.Optional[str] = None, config_name: t.Optional[str] = None
    ) -> t.List[str]:
        return self._most_suitable_rule(folder).enable_build_targets(default_sdkconfig_target, config_name)

    def enable_test_targets(
        self, folder: str, default_sdkconfig_target: t.Optional[str] = None, config_name: t.Optional[str] = None
    ) -> t.List[str]:
        return self._most_suitable_rule(folder).enable_test_targets(default_sdkconfig_target, config_name)
```

File: idf_build_apps/manifest/manifest.py (hunk 1 above)
Comment (horw) [reply idf_build_apps/manifest/manifest.py:120]: I added a few more lines to the code, but typing right now

File: idf_build_apps/manifest/manifest.py [hunk 4]
```diff
@@ -46,6 +46,19 @@ def get_value(self, target: str, config_name: str) -> t.Any:
         return self.stmt.get_value(target, config_name)
 
 
+class SwitchClause:
+    def __init__(self, ifs: t.List[IfClause], values: t.List[t.List[str]], default: t.List[str]) -> None:
```
[the remark this answers is not in the document]
Comment (horw) [reply idf_build_apps/manifest/manifest.py:50]: Good name

File: docs/manifest.md [hunk 5]
```diff
@@ -51,6 +51,32 @@ All operators are binary operators. For more than two operands, you may use the
 - `A == 1 or (B == 2 and C in [1,2,3])`
 - `(A == 1 and B == 2) or (C not in ["3", "4", 5])`
 
+## `switch` Clauses
```
[the remark this answers is not in the document]
Comment (horw) [reply docs/manifest.md:54]: Moved

File: idf_build_apps/manifest/manifest.py (hunk 2 above)
Comment (horw) [reply idf_build_apps/manifest/manifest.py]: BTW, I renamed it to "target" in two functions. If you also think it's better to keep the name "default_sdkconfig_target," I can revert the changes.

Comment (hfudev) [reply idf_build_apps/manifest/manifest.py]: oh, sorry for misleading you. my bad :(

yes you're right. I was comparing it to some other functions. could you help revert them to `default_sdkconfig_target` instead? Thanks a lot.

**Review by hfudev — APPROVED** [review]
Cool. kudos to the nice test cases. LGTM!

**Pull request merged**